fedimint-bin-tests/src/main.rs

```diff
@@ -958,14 +958,97 @@ async fn reconnect_test(dev_fed: DevFed, process_mgr: &ProcessManager) -> Result
     fed.kill_server(2).await?;
     fed.kill_server(3).await?;
 
-    fed.start_server(process_mgr, 1).await?;
-    fed.start_server(process_mgr, 2).await?;
-    fed.start_server(process_mgr, 3).await?;
+    fed.start_server(process_mgr, 1, None).await?;
+    fed.start_server(process_mgr, 2, None).await?;
+    fed.start_server(process_mgr, 3, None).await?;
     fed.await_all_peers().await?;
     info!("fm success: reconnect-test");
     Ok(())
 }
 
+async fn upgrade_test(dev_fed: DevFed, process_mgr: &ProcessManager) -> Result<()> {
+    #[allow(unused_variables)]
+    let DevFed {
+        bitcoind,
+        cln,
+        lnd,
+        mut fed,
+        gw_cln,
+        gw_lnd,
+        electrs,
+        esplora,
+    } = dev_fed;
+
+    let data_dir = env::var("FM_DATA_DIR")?;
+
+    fed.await_block_sync().await?;
+
+    info!("### Signal upgrade peer0");
+    cmd!(
+        fed,
+        "signal-upgrade",
+        "--salt-path={data_dir}/server-0/private.salt",
+        "--our-id=0"
+    )
+    .env("FM_PASSWORD", "pass0")
+    .run()
+    .await?;
+
+    info!("Signal upgrade peer1");
+    cmd!(
+        fed,
+        "signal-upgrade",
+        "--salt-path={data_dir}/server-1/private.salt",
+        "--our-id=1"
+    )
+    .env("FM_PASSWORD", "pass1")
+    .run()
+    .await?;
+
+    bitcoind.mine_blocks(1).await?;
+    fed.await_block_sync().await?;
+
+    let epoch_count = cmd!(fed, "epoch-count").out_json().await?["count"]
+        .as_u64()
+        .unwrap();
+    let upgrade_epoch = epoch_count + 1;
+
+    info!("Signal upgrade peer2");
+    cmd!(
+        &fed,
+        "signal-upgrade",
+        "--salt-path={data_dir}/server-2/private.salt",
+        "--our-id=2"
+    )
+    .env("FM_PASSWORD", "pass2")
+    .run()
+    .await?;
+
+    info!("Wait for server shutdowns");
+    fed.kill_server(0).await?;
+    fed.kill_server(1).await?;
+    fed.kill_server(2).await?;
+    fed.kill_server(3).await?;
+
+    info!("Restart federation");
+    fed.start_server(process_mgr, 0, Some(upgrade_epoch))
+        .await?;
+    fed.start_server(process_mgr, 1, Some(upgrade_epoch))
+        .await?;
+    fed.start_server(process_mgr, 2, Some(upgrade_epoch))
+        .await?;
+    fed.start_server(process_mgr, 3, Some(upgrade_epoch))
+        .await?;
+
+    info!("Federation started");
+
+    bitcoind.mine_blocks(1).await?;
+    fed.await_block_sync().await?;
+    fed.await_all_peers().await?;
```

Comment:
Having some problem with `await_all_peers`
`command: /home/dhruv/development/oss/forked/fedimint/target/debug/fedimint-cli --data-dir=/tmp/nix-shell.UEIDMP/fm-EhYd/cfg api /module/2/block_height
stdout:

stderr:
2023-04-19T14:06:58.194586Z ERROR net::api: unable to connect to server err=Networking or low-level protocol error: Error when opening the TCP socket: Connection refused (os error 111)`

Reply:
The Federation seems to be running, i'm not sure why this command is failing. Any possible reason for this failure?

Reply:
I was killing `fedimintd`, instead I should have waited the it's shutdown.